File: 26_rxjava2/src/main/java/com/huoergai/rxjava2/MainActivity.java
package com.huoergai.rxjava2;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import java.util.List;
import java.util.Locale;
import java.util.concurrent.TimeUnit;

import hu.akarnokd.rxjava3.retrofit.RxJava3CallAdapterFactory;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Observer;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.core.SingleObserver;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.functions.Function;
import io.reactivex.rxjava3.schedulers.Schedulers;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class MainActivity extends AppCompatActivity {

    private TextView tv;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        tv = findViewById(R.id.tv);
    }

    public void onClick(View v) {
        // fetchRepos();
        // rxjavaTest1();
        rxjavaTest2();
    }

    private void fetchRepos() {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://api.github.com/")
                .addConverterFactory(GsonConverterFactory.create())
                // .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addCallAdapterFactory(RxJava3CallAdapterFactory.create())
                .build();
        NetApi netApi = retrofit.create(NetApi.class);
        netApi.getRepos("huoergai")
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new SingleObserver<List<Repo>>() {
                    @Override
                    public void onSubscribe(@NonNull Disposable d) {
                        tv.setText(R.string.loading);
                    }

                    @Override
                    public void onSuccess(@NonNull List<Repo> repos) {
                        tv.setText(repos.get(2).full_name);
                    }

                    @Override
                    public void onError(@NonNull Throwable e) {
                        tv.setText(e.getMessage());
                    }
                });
    }

    private void rxjavaTest1() {
        Single<String> stringSingle = Single.just("test")
                .map(new Function<String, String>() {
                    @Override
                    public String apply(String s) throws Throwable {
                        return "mapped+" + s;
                    }
                });
        stringSingle.subscribe(new SingleObserver<String>() {
            @Override
            public void onSubscribe(Disposable d) {
            }

            @Override
            public void onSuccess(String s) {
                tv.setText(s);
            }

            @Override
            public void onError(Throwable e) {

            }
        });
    }

    private void rxjavaTest2() {
        Observable.interval(1, TimeUnit.SECONDS)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<Long>() {
                    @Override
                    public void onSubscribe(@NonNull Disposable d) {
                        tv.setText(R.string.start);
                    }

                    @Override
                    public void onNext(@NonNull Long aLong) {
                        tv.setText(String.format(Locale.CHINA, "%ds", aLong));
                    }

                    @Override
                    public void onError(@NonNull Throwable e) {
                        tv.setText(R.string.error);
                    }

                    @Override
                    public void onComplete() {
                        tv.setText(R.string.complete);
                    }
                });

    }

}

File: 17_touch_event/src/main/java/com/huoergai/touch_event/MainActivity.kt
package com.huoergai.touch_event

import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.snackbar.Snackbar

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        /*
         * 1.触摸事件的消费顺序是：从上层子 View 往下到根 View
         * 2.一旦在中间被消费(在 OnTouchEvent 的 ACTION_DOWN 事件中返回了 true), 则不再往下传递,
         *   该同一组或同一序列的后续事件也不再向下传递。
         * 3.getAction() 与 getActionMasked(): 后者有对多点触控的支持,除了能获取到事件类型，还能获取手指 index 等信息
         * 4.ViewGroup onInterceptTouchEvent() 负责是否拦截子 View 的事件, 拦截后则其 onTouchEvent() 负责处理该事件
         * 5.如果 ViewGroup 拦截了事件, 需要在拦截时记录下按下的数据,以供后面消费事件使用
         *
         * 触摸事件传递流程简化
         * View.dispatchTouchEvent():
         * public boolean dispatchTouchEvent(){
         *      return onTouchEvent();
         * }
         *
         * ViewGroup.dispatchTouchEvent()
         * public boolean ViewGroup.dispatchTouchEvent(){
         *      boolean result;
         *      if(interceptTouchEvent()){
         *          result = onTouchEvent();
         *      }else{
         *          result = 子 View 的 dispatchTouchEvent()
         *      }
         *      return result
         * }
         *
         */



    }

    fun onClick(view: View) {
        when (view.id) {
            R.id.tl -> {
                Snackbar.make(view, "TouchLayout ...", Snackbar.LENGTH_SHORT).show()
            }
            else -> {
                Snackbar.make(view, "TouchView ...", Snackbar.LENGTH_SHORT).show()
            }
        }
    }
}
